Reduce build time and fix timeouts on Travis

nilsdeppe opened the pull request:
## Proposed changes

- Improves build time slightly (~7-8%). I'll push this down more in a future PR but that's already a useful amount, especially when building with GCC.
- Changes TravisCI to build the `Evolve*` executables one at a time to avoid timeouts on release builds.

I'm treating these as "bug fixes" since they aren't really new features.

Marking as priority since it'll reduce developer time spent building executables and reduce the number of times we have to refresh TravisCI builds.

### Types of changes:

- [x] Bugfix

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [x] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Fix .cpp include in executables

On my system reduces GCC6 build time by ~7%

nilsdeppe committed:
Prevent #include'ing cpp files in cpp, hpp, and tpp files

nilsdeppe committed:
Speed up tmpl::list_difference by factor of ~2

nilsdeppe committed:
Use static_assert instead of Requires in DataBox constructor

nilsdeppe committed:
Limit more assertions to debug mode DataBox

Helps reduce compile time in release mode

nilsdeppe committed:
Build executables in serial on TravisCI

kidder commented:
@wthrowe please look at

wthrowe commented on tools/FileTestDefs.sh:
```diff
@@ -414,6 +414,26 @@ namespace_details_test() {
 
+# Check for .cpp includes in headers and tpp files
+cpp_includes_in_includibles() {
+    is_includible "$1" && staged_grep -q "include .*\.cpp" $1
```
Is there a reason to allow this in cpp files?

nilsdeppe replied:
It seems "less wrong" but I think with `tpp` files we have a design that should not require us to ever include a cpp, right?

wthrowe replied:
Yeah.  Currently, there is one such include in `tests/Unit/DataStructures/Test_OrientVariablesOnSlice.cpp`, which is a bug, and a bunch of includes of `Parallel/CharmMain.cpp`, which would have to be dealt with.

So I guess our choices are: (1) rename CharmMain.cpp to a tpp or something, (2) whitelist CharmMain.cpp includes, or (3) leave this check as is for now and decide what to do with the cpp case later because it will get Travis fixed faster.  I'm fine with any of these.

nilsdeppe replied:
I think for the `ChairmMain.cpp` I'll go with `tpp` for now. While not "techincally consistent" (it doesn't contain template definitions) it should be used in basically the same way.

nilsdeppe committed:
Rename CharmMain.cpp to CharmMain.tpp

This is to avoid including cpp files anywhere since including cpp files is
somewhat unusual. This also allows us to easily have a test that checks for cpp
includes in files.

kidder commented:
@wthrowe please look at

wthrowe commented on tools/FileTestDefs.sh:
```diff
@@ -413,6 +413,29 @@ namespace_details_test() {
+    is_c++ "$1" && staged_grep -q "include .*\.cpp" $1
+}
+cpp_includes_in_includibles_report() {
+    echo "Found cpp files included in header or tpp files."
```
Check name, check comment, error message, and commit message need updating.

wthrowe commented on tools/FileTestDefs.sh:
```diff
@@ -413,6 +413,29 @@ namespace_details_test() {
 
+# Check for .cpp includes in headers and tpp files
+cpp_includes_in_includibles() {
+    is_c++ "$1" && staged_grep -q "include .*\.cpp" $1
```
Quote second `$1`.

nilsdeppe replied:
For my own understanding, what is the difference in meaning between the quoted and unquoted version?

wthrowe replied:
The unquoted version does word splitting, so if `$1` is `Hello world` then `func "$1"` is the same as `func "Hello world"` and `func $1` is the same as `func "Hello" "world"`.  It doesn't really matter here since I'm sure lots of other things will be unhappy if we have filenames containing whitespace, but this file has pretty good bash style overall so I'm being slightly picky about it.

nilsdeppe replied:
Ah okay, that makes sense :) I appreciate you teaching me the bash ways :)

nilsdeppe commented:
Thanks @wthrowe! Should be done :)
